### PR TITLE
Added PWA

### LINKED ISSUE #801: Turn Telescope into a Progressive Web App (PWA)
**What would you like to be added**:
We briefly talked about this a long time ago and I think our Gatsby website is ready to be a PWA now

**Why would you like this to be added**:
Our frontend has been incredibly improved the last few days for mobile. PWA is the next step

### PR DESCRIPTION
**Pull request by tonyvugithub**
## Issue This PR Addresses
Fixed #801 

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description
This PR turns our Telescope into a Progressive Web App (PWA) to improve user experience on mobile. 
Reference: https://github.com/vercel/next.js/tree/canary/examples/progressive-web-app

Choose `Install app` on Chrome

## Checklist

- [ ] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

**Comment by humphd** [comment]
@tonyvugithub please add more details to this beyond "Ongoing process..."

**Comment by humphd** [comment]
If this is ready for review, remove Draft and let me know.

**Comment by tonyvugithub** [comment]
@humphd : I am still adding some screenshots and it should be ready for reviewing

**Comment by chrispinkney** [comment]
Wow, nice. I just tested the deployment on my phone (using Firefox) and it works really well! There are some issues (like logging in not working, or search being broken or about page 404ing) but the deployment had the same issues so this is not a PWA problem.

Here's some screenshots I took:

Screenshot_20210326-123615_Firefox

Screenshot_20210326-123622_One UI Home

Screenshot_20210326-123642_One UI Home

Screenshot_20210326-123650_Firefox

File: src/web/next.config.js
Comment (chrispinkney) [inline src/web/next.config.js:57]: This is why the about page wasn't rendering (noted in my screenshot comment), I would love to see the solution to this

Comment (tonyvugithub) [reply src/web/next.config.js]: @chrispinkney : I am working on it

**Comment by tonyvugithub** [comment]
@humphd , @chrispinkney The conflict problem is bigger than I thought. So far none of my attempt has worked. Should we consider upgrade `About` to a component?

**Comment by humphd** [comment]
I feel like we're hitting a lot of bugs that relate to About, and not actually doing a ton with it.  I'm fine to alter how it works for sure.

**Comment by tonyvugithub** [comment]
@chrispinkney : Are you ok with converting the About.mdx to a page component?

**Comment by chrispinkney** [comment]
> 
> 
> @chrispinkney : Are you ok with converting the About.mdx to a page component?

I'm okay with it happening, but I'd rather stay focused on the users microservice right now if that's okay.

Regardless it's really too bad that we couldn't get both kids to get along 😞

**Comment by tonyvugithub** [comment]
@chrispinkney already did and i came across weird bug.

**Comment by tonyvugithub** [comment]
@chrispinkney : I tried both `next-compose` and `next-compose-plugins`

**Comment by humphd** [comment]
> @chrispinkney : I tried both next-compose and next-compose-plugins

Can you say more about what you did, and what happened?

**Comment by tonyvugithub** [comment]
@humphd : for `next-compose`, I was able to make the site a PWA, but the `withMDX` plugin did not kick in resulting in it being not found when access. That's why I was thinking to upgrade `About` to a page component and the problem should be solved. I tried to find any recommendation to use `withMDX` and `withPWA` but no luck. 
For `next-compose-plugins`, I could not build it. The `withPWA` was loaded probably but when it trying to compile the `About` page, it failed.
image

**Comment by tonyvugithub** [comment]
@humphd I am doing another attempt to ensure that I did not make a mistake.

**Review by birtony — COMMENTED** [review]
Tried it on an iPhone, looks pretty good and the About page works. However, the icon is not detected correctly for some reason (it uses the banner image instead of the icon from `/src/web/public` folder). About page footer is hidden behind the bottom navbar.  The search did not work. App worked in airplane mode, so when we have our caching policies improved it should be very convenient. See screenshots below:
IMG_3184
IMG_3185
IMG_3186
IMG_3187
IMG_3188
IMG_3189
IMG_3180
IMG_3181
IMG_3183

Also, I've noticed our search results fail to load on PR previews, which is not specific to this PR only.

**Comment by tonyvugithub** [comment]
@humphd @chrispinkney @birtony : Hi, a bit of update from this PR. It seems that I got a bit unlucky with my attempt, right when I thought I would figure it out and I literally spent 3 hours on making different attempts but the problem might have come from upstream of one of the dependencies of `next-pwa`. 
https://github.com/shadowwalker/next-pwa/issues/177

**Comment by huynguyez** [comment]
cc @tonyvugithub  It's been a while and we wanted to know what is the progress on this PR?

**Comment by tonyvugithub** [comment]
@huynguyez : Its no longer blocked. Ready for reviewing.

**Comment by tonyvugithub** [comment]
@huynguyez its ready to review. I replied to you yesterday.

**Review by birtony — COMMENTED** [review]
@tonyvugithub is there any update to the iOS side of things?

**Comment by tonyvugithub** [comment]
@birtony : I don't think you have the 'install' prompt as in Android but you can use the 'Add to Home Screen' option. For now it is how it should be on iOS.
Reference Link: https://mobilesyrup.com/2020/05/24/how-install-progressive-web-app-pwa-android-ios-pc-mac/

**Comment by birtony** [comment]
> @birtony : I don't think you have the 'install' prompt as in Android but you can use the 'Add to Home Screen' option in Safari. For now it is how it should be on iOS.
> 
> Reference Links: 
> 
> https://mobilesyrup.com/2020/05/24/how-install-progressive-web-app-pwa-android-ios-pc-mac/
> 
> https://brainhub.eu/library/pwa-on-ios/

@tonyvugithub okay, I see. However, I think we are at least missing a proper icon on iOS at the moment:
image

**Comment by tonyvugithub** [comment]
@birtony potentially it is because iOS did not pick up the manifest.json file but I will look deeper into this.

**Review by chrispinkney — DISMISSED** [review]
Just wanted to chime in and I'm unsure if an approval is the right way to do it but, why not?

I just tested it again, everything (light/dark mode, search, about page) is still working well on my Galaxy S8 w/ Firefox 87.0.0-rc.1. I'd test it on more devices but this is all I have.

I'm just not sure how else to test this. If anyone wants any additional/more advanced testing on Android feel free to ping me. LMK!

PS: I don't have the icon issue like Anton does, mine still looks as it does in the screenshot I posted above.

**Comment by birtony** [comment]
> 
> 
> @birtony potentially it is because iOS did not pick up the manifest.json file but I will look deeper into this.

Hey @tonyvugithub, have you had a chance to look into the iOS icon bug?

**Comment by tonyvugithub** [comment]
@birtony : Yes I did a bit but did not get a chance to implement it. Basically you need an `apple-touch-icon`. https://web.dev/codelab-apple-touch-icon/

**Comment by tonyvugithub** [comment]
@birtony : I updated the `apple=touch-icon`. Can you check if it shows the icon on iOS now?

**Comment by birtony** [comment]
Just checked it in Safari on iOS in private mode and it is still not picking it @tonyvugithub

**Comment by tonyvugithub** [comment]
@birtony : It should be working now. I typed `app-touch-icon` instead of `apple-touch-icon`, that was why It did not pick up.

**Comment by PedroFonsecaDEV** [comment]
It's working.

**Review by PedroFonsecaDEV — DISMISSED** [review]

**Review by huynguyez — COMMENTED** [review]
Tested on Chrome (Mac) and working now, but very small issue, is there anyway to change this icon shape?
Screen Shot 2021-04-14 at 8 01 22 PM

**Comment by tonyvugithub** [comment]
@huynguyez : I will try to look into it today.

**Comment by tonyvugithub** [comment]
@huynguyez : Actually it is the icon that has the square background https://github.com/Seneca-CDOT/telescope/blob/master/src/web/public/logoImages/manifest-icon-192.png. 
If you want a rounded icon then @PedroFonsecaDEV needs to create one for me.

**Review by chrispinkney — DISMISSED** [review]

**Comment by PedroFonsecaDEV** [comment]
@huynguyez
IMG_6953292F943A-1
The icon looks fine on my phone. We can investigate how to make it looks prettier on a Mac, but this is going to be in a follow-up. 

Congrats @tonyvugithub, let's land this.

**Comment by humphd** [comment]
Based on conversation on Slack, we agreed that we are going to wait on the response from the next PWA maintainer about how the update path works.  If things will be OK, and not require custom UI from us, this is probably good to land.

**Comment by tonyvugithub** [comment]
@humphd He should reply me next couple of days as he still actively maintain the repo.

**Comment by tonyvugithub** [comment]
@PedroFonsecaDEV @chrispinkney @humphd : The `next-pwa` creator got back to me and he confirmed that if we changed prod it will update the cache in the PWA.
image

File: src/web/.babelrc
Comment (humphd) [inline src/web/.babelrc:2]: Can we not use the `.bablerc` that we have in the root?  Why do we need a second one?

Comment (tonyvugithub) [reply src/web/.babelrc]: @humphd Yes we can. I made change to it and got rid of the second babel file

Comment (tonyvugithub) [reply src/web/.babelrc]: @humphd : Actually when I got rid of the babel file inside /web. It failed the unit test.
image

File: src/web/next.config.js
Comment (humphd) [inline src/web/next.config.js:78]: Let's do it the same way they do it in https://www.npmjs.com/package/next-compose-plugins#nextconfiguration with a `nextConfig` object above, so it's clear what is the base config, and what is a plugin config.

File: src/web/package.json
Comment (humphd) [inline src/web/package.json:12]: There are more changes to this file than I'd expect.  Can you confirm they are all needed?

File: src/web/public/manifest.json
Comment (humphd) [inline src/web/public/manifest.json:12]: We have a lot more icon sizes than just the 192x192 one, why only this one?

**Review by humphd — CHANGES_REQUESTED** [review]

File: src/web/.babelrc
Comment (tonyvugithub) [reply src/web/.babelrc]: this is solved

File: src/web/public/manifest.json
Comment (tonyvugithub) [reply src/web/public/manifest.json:12]: I can add a few more. I include one for simplicity initally

Comment (humphd) [reply src/web/public/manifest.json:12]: That would be good, yes.  Let's do it either in this PR, or file an issue and we can follow-up with it later.

File: src/web/package.json
Comment (tonyvugithub) [reply src/web/package.json:12]: A few I can confirm. I was wondering myself too. I'm cleaning it up now.

File: src/web/src/pages/_document.tsx
Comment (humphd) [inline src/web/src/pages/_document.tsx:32]: We have a hook that manages the icon and adds a red dot when there are new posts.  We should make sure this won't break anything with what we do in https://github.com/Seneca-CDOT/telescope/blob/master/src/web/src/hooks/use-favicon-badge.ts#L14.

Comment (tonyvugithub) [reply src/web/src/pages/_document.tsx:32]: @humphd I think I will get rid of these links and only leave the `apple-touch-icon` link

File: src/web/package.json
Comment (humphd) [inline src/web/package.json:40]: Where is this used?

**Review by humphd — APPROVED** [review]
I think we can try this, and iterate on it later.  One thing we could do that I notice is add caching for our posts api, so they get saved at runtime for offline access later.

File: src/web/package.json
Comment (tonyvugithub) [reply src/web/package.json:40]: When I implemented `next-pwa` and using webpack 5, there was an error with webpack saying that `terser-webpack-plugin` cannot be found. So I installed it as a devDep and the error gone.

Comment (tonyvugithub) [reply src/web/package.json:40]: It is for sth used inside `next-pwa`

**Comment by tonyvugithub** [comment]
@humphd That was sth I have in mind too as I know the image is cached but not the posts. However, it would require more research into the matter. I can always do a follow up in the future.

**Review by chrispinkney — APPROVED** [review]
Woohoo!

**Pull request merged**
